### ISSUE TITLE
Add node for apply script to parent

### ISSUE THREAD

**Issue by mageowl** [godotengine/godot-proposals#8934]
### Describe the project you are working on

Roguelike with collectable upgrades that modify the player controller.

### Describe the problem or limitation you are having in your project

It is currently hard to have a script inject code into a different node's script, without writing your own boilerplate code. This is a very useful feature for games with upgrades that don't just change a number statistic, but make changes to the player controller, for example blocking any damage that is from fire enemies, or reducing knockback by 50%.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be useful to be able to have a node called something like `ScriptNode` that when a script is attached to it, it adds it to the parent node, instead of applying it to itself. They could be stacked, and would apply in the order of the node hierarchy.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For example, say the player has a function `take_damage`.
```gdscript
class_name Player extends CharacterBody2D

func take_damage(amount, type):
    hp -= type
```
Then, you pick up an upgrade that blocks all incoming fire damage. A `ScriptNode` is added to the player with the following script:
```gdscript
extends Player

func take_damage(amount, type):
    if type != "fire":
        super(amount, type)
```
The `ScriptNode` applies the upgrade script to the player, and anytime the `take_damage` function is called, if the damage type is not `"fire"`, then the `super` is called, running the `take_damage` function in the `Player` script. If the damage _is_ fire, nothing happens. You could then have ones for `"lightning"`, `"acid"`, and `"drowning"`, and they could all be active at once, with each `super` calling the `ScriptNode` higher up on the list's `take_damage` function.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You can use signals and functions like so:

```gdscript
# Player
signal mod_take_damage(args: Dictionary)

func take_damage(type, amount):
    var args = {"type": type, "amount": amount}
    mod_take_damage.emit(args)
    hp -= args.amount

# Upgrade
func take_damage(args): # Connected to the player's mod_take_damage signal
    if args.type == "fire":
        args.amount = 0
```

But it is more cumbersome, and there is no way to completely cancel the event.

### Is there a reason why this should be core and not an add-on in the asset library?

This is extremely flexible and can be used in many different applications.

**Comment by Mickeon** [comment]
The whole structure of this code is kind of an anti-pattern. I'm not convinced adding a whole node that encourages this would benefit many. That's not to mention this **ScriptNode**'s name and behavior is not really intuitive by itself. 

There's many different ways to go about alternatives that are more "proper" while still keeping the code very dynamic, as is the main issue in the proposal. I'm not sure where to start.

**Comment by mageowl** [comment]
I think that being able to apply traits at runtime would solve this issue.

**Comment by dalexeev** [comment]
No, traits do not imply runtime application and also do not support multi-adding of a method with the same name from different traits (in this case you must explicitly resolve the conflict).

The same applies to this proposal: it is not obvious in what order the modifiers should be called, whether a modifier can stop the chain, whether a modifier can call parent class method (that's why `super` in your example is not good, that would require a different keyword).

GDScript doesn't allow for this kind of function code patching at runtime, and you can't replace a method with a wrapper like in JavaScript (you could emulate this with `Callable`s, but that's probably a bad idea).

Also, `Node` is not a base class. Yes, we have constructs for `Node` in the language (like `@onready` and `$Node`), but function is a more fundamental concept than `Node`.

**Comment by Mickeon** [comment]
I will repeat, there's many ways to go about this issue that are not just straightforward but more structurally sound. looking up "Node as Components" may shed some light on how to make use of Nodes very dynamically.

One _other_ way to go about it, specifically tailored to the example in the proposal, is to make use of RefCounted/Resources.
The goal is to make use of another object to swap out logic as desired.

Create some sort of "Defense" object with the following function:
```GDScript
func should_resist_damage(amount, type):
    return type == "fire"
```

Add a `defense` property to the **Player** script. When you collect a powerup, assign a new RefCounted with the script above attached. Then, the following bit of logic:
```GDScript
func take_damage(amount, type):
    if defense != null and defense.should_resist_damage(amount, type):
        return # Do nothing.

    hp -= type
```

This is not an uncommon pattern. It's a form of dependency injection. Perhaps it may look ugly because I do not know the full context, and as such I cannot suggest a whole set of names and methods for this class.
